dipy/denoise/gibbs.py

```diff
@@ -0,0 +1,310 @@
+
+import numpy as np
+
+
+def image_tv(x, a=0, nn=3):
+    """ Computes total variation (TV) of matrix x along axis a in two
+    directions.
+
+    Parameters
+    ----------
+    x : 2D ndarray
+        matrix x
+    a : int (0 or 1)
+        Axis along which TV will be calculated. Default a is set to 0.
+    nn : int
+        Number of points to be included in TV calculation.
+
+    Returns
+    -------
+    PTV : 2D ndarray
+        Total variation calculated from the right neighbors of each point
+    NTV : 2D ndarray
+        Total variation calculated from the left neighbors of each point
+
+    Note
+    ----
+    This function was created to deal with gibbs artefacts of MR images.
+    Assuming that MR images are reconstructed from estimates of their Fourier
+    expansion coefficients, during TV calculation matrix x can taken as and
+    periodic signal. In this way NTV values on the image left boundary is
+    computed using the time series values on the right boundary and vice versa.
+    """
+    if a:
+        xs = x.copy()
+    else:
+        xs = x.T.copy()
+
+    # Add copies of the data so that data extreme points are also analysed
+    xs = np.concatenate((xs[:, (-nn-1):], xs, xs[:, 0:(nn+1)]), axis=1)
+
+    PTV = np.absolute(xs[:, (nn+1):(-nn-1)] -
+                      xs[:, (nn+2):(-nn)])
+    NTV = np.absolute(xs[:, (nn+1):(-nn-1)] -
+                      xs[:, nn:(-nn-2)])
+    for n in range(1, nn):
+        PTV = PTV + np.absolute(xs[:, (nn+1+n):(-nn-1+n)] -
+                                xs[:, (nn+2+n):(-nn+n)])
+        NTV = NTV + np.absolute(xs[:, (nn+1-n):(-nn-1-n)] -
+                                xs[:, (nn-n):(-nn-2-n)])
+
+    if a:
+        return PTV, NTV
+    else:
+        return PTV.T, NTV.T
+
+
+def gibbs_removal_1d(x, a=0, nn=3):
+    """ Decreases gibbs ringing along an axis 'a' using fourier sub-shifts.
+
+    Parameters
+    ----------
+    x : 2D ndarray
+        Matrix x.
+    a : int (0 or 1)
+        Axis along which gibbs oscilations will be reduced. Default a is set
+        to 0 (i.e. gibbs are reduce along axis y).
+    nn : int, optional
+        Number of neighbour points to access local TV (see note). Default is
+        set to 3.
+
+    Returns
+    -------
+    xc : 2D ndarray
+        Matrix with gibbs oscilantions reduced along an axis 'a'.
+
+    Note
+    ----
+    This function decreases the effects of gibbs oscilations based on the
+    analysis of local total variation (TV). Although artefact correction is
+    done based on two adjanced points for each voxel, total variation should be
+    accessed in a larger range of neigbors. If you want to adjust the number
+    of the neigbors to be considered in TV calculation please change
+    parameter nn.
+    """
+    ssamp = np.linspace(0.02, 0.9, num=45)
+
+    if a:
+        xs = x.copy()
+    else:
+        xs = x.T.copy()
+
+    # TV for shift zero (baseline)
+    TVR, TVL = image_tv(xs, a=1, nn=nn)
```

Comment:
Should the axis variable `a` really always be set to `1`? Can't you use the input to avoid the if statement in L88?